src/script/script.hpp

```diff
@@ -0,0 +1,42 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+ /** @file script.hpp Base functions for all Scripts. */
+
+#ifndef SCRIPT_HPP
+#define SCRIPT_HPP
+
+#include "../ai/ai.hpp"
+#include "../game/game.hpp"
+
+/**
+ * Main Script class. Contains functions needed to handle Script Events.
+ */
+class Script {
+public:
+	/**
+	 * Queue two new events, one for an AI, the other for the Game Script.
+	 * @param company The company receiving the event.
+	 */
+	template <class ScriptEventType, typename ... Args>
+	static void NewEvent(CompanyID company, Args ... args) {
+		AI::NewEvent(company, new ScriptEventType(args...));
+		Game::NewEvent(new ScriptEventType(args...));
+	}
+
+	/**
+	 * Broadcast a new event to all active AIs, and to the Game Script.
+	 * @param skip_company The company to skip broadcasting for, or MAX_COMPANIES to skip none.
+	 */
+	template <class ScriptEventType, typename ... Args>
+	static void BroadcastNewEvent(CompanyID skip_company, Args ... args) {
+		AI::BroadcastNewEvent(new ScriptEventType(args...), skip_company);
+		Game::NewEvent(new ScriptEventType(args...));
+	}
+};
```

Comment:
I really like these new helpers. It would be nice to introduce them first (including changing the existing ones), and then as you add the different events change those calls from AI::NewEvent to the new helper. Now the order is a bit weird, since you're adding Game::NewEvent calls to replace them a few commits later.